src/utilities/convex_hull.zig

```diff
@@ -185,10 +187,56 @@ pub const ConvexHull = struct {
         return LinearFunction{ .slope = min_slope, .intercept = min_intercept };
     }
 
+    /// Merges another `other: ConvexHull` into the current `self: ConvexHull` in place. This
+    /// operation modifies the `self` object by appending points from the `other` convex hull.
+    /// The function assumes both convex hulls are valid and sorted. If appending to the lower
+    /// or upper hull fails due to allocation issues, the error is propagated.
+    pub fn merge(self: *ConvexHull, other: *ConvexHull) !void {
+        // Special case: if `other` has only one point, directly add it to self's hulls
+        if (other.upper_hull.items.len == 1 and other.lower_hull.items.len == 1) {
+            try self.addPoint(other.upper_hull.items[0]);
+            return;
+        }
+
+        // Special case: if `self` has only one point, transfer points from `other` directly
+        if (self.upper_hull.items.len == 1 and self.lower_hull.items.len == 1) {
+            const other_points = try other.getAllPointsSorted(); // Handle the error union
+            defer other_points.deinit(); // Ensure the allocated list is freed
+            for (other_points.items) |point| {
+                try self.addPoint(point); // Add each point to `self`
+            }
+            return;
+        }
+        // Find tangent between the upper hulls.
+        const upper_tangent = try findTangent(
+            &self.upper_hull,
+            &other.upper_hull,
+            .upperHull,
+        );
+        // Find tangent between the lower hulls.
+        const lower_tangent = try findTangent(
+            &self.lower_hull,
+            &other.lower_hull,
+            .lowerHull,
+        );
+
+        // Add points to `self`'s upper hull starting from the right-most point
+        // of the `other`s upper hull's tangent.
+        for (upper_tangent.hull_two_idx..other.upper_hull.items.len) |i| {
+            try self.upper_hull.append(other.upper_hull.items[i]);
+        }
+
+        // Add points to `self`'s lower hull starting from the right-most point
+        // of the `other`s lower hull's tangent.
+        for (lower_tangent.hull_two_idx..other.lower_hull.items.len) |i| {
+            try self.lower_hull.append(other.lower_hull.items[i]);
+        }
+    }
+
     /// Helper function to get all points as an `ArrayList(DiscretePoint)` from the ConvexHull.
     /// If the `ConvexHull` is empty, an empty ArrayList is returned.
     fn getAllPoints(self: *ConvexHull) !ArrayList(DiscretePoint) {
-        var hull: ArrayList(DiscretePoint) = ArrayList(DiscretePoint).initCapacity(
+        var hull: ArrayList(DiscretePoint) = try ArrayList(DiscretePoint).initCapacity(
             self.allocator,
             self.lower_hull.items.len + self.upper_hull.items.len - 2,
```

Comment:
Maybe add a comment why space for `- 2` `DiscretePoint`s is allocated.

Reply:
Sure, answer related to https://github.com/cmcuza/TerseTS/pull/57#discussion_r1864272784